micro-rdk/src/common/system.rs

```diff
@@ -0,0 +1,96 @@
+use std::fmt::Display;
+use std::sync::{Arc, LazyLock};
+use std::time::Duration;
+
+#[cfg(not(feature = "esp32"))]
+use async_io::Timer;
+use async_lock::Mutex as AsyncMutex;
+
+use super::app_client::{AppClient, PeriodicAppClientTask};
+use super::log::LogUploadTask;
+use super::runtime::terminate;
+
+pub(crate) static SHUTDOWN_EVENT: LazyLock<Arc<AsyncMutex<Option<SystemEvent>>>> =
+    LazyLock::new(|| Arc::new(AsyncMutex::new(None)));
+
+pub(crate) enum SystemEvent {
+    Restart,
+    #[allow(dead_code)]
+    DeepSleep(Option<Duration>),
+}
+
+impl Display for SystemEvent {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Self::Restart => "restart".to_string(),
+                Self::DeepSleep(dur) => dur.as_ref().map_or("enter deep sleep".to_string(), |d| {
+                    format!("enter deep sleep for {} microseconds", d.as_micros())
+                }),
+            }
+        )
+    }
+}
+
+pub(crate) async fn send_system_change(event: SystemEvent) {
+    log::info!("received call to {}", event);
+    let _ = SHUTDOWN_EVENT.lock().await.insert(event);
```

Comment:
if we have multiple event they are going to be overwritten no?

Reply:
I was thinking that we actually want to respect the latest requested event?

Reply:
never mind, that decision's ridiculous. I'll fix this